app/controllers/file_manager_controller.rb

```diff
@@ -0,0 +1,234 @@
+      if parent == BASE_DIRECTORY
+        flash[:error] = "Unable to rename courses in the root directory."
+      else
+        dir_name = File.dirname(params[:relative_path])
+
+        if params[:new_name].empty? || params[:new_name].nil?
+          raise ArgumentError, "New name not provided"
+        end
+
+        unless params[:new_name].match(/^[a-zA-Z0-9_-]+(\.[a-zA-Z0-9_-]+)?$/)
+          raise ArgumentError, "Invalid characters. Only letters,
+        numbers, underscores, and hyphens are allowed."
+        end
+
+        new_path = safe_expand_path("#{dir_name}/#{params[:new_name]}")
+        parent = new_path.split[0..-2].join('/')
+
+        raise ArgumentError, "A file with that name already exists" if File.exist?(new_path)
+
+        FileUtils.mkdir_p(parent)
+        FileUtils.mv(absolute_path, new_path)
+        flash[:success] = "Successfully renamed file to #{params[:new_name]}"
+      end
+    end
+  rescue ArgumentError => e
+    flash[:error] = e.message
+  end
+
+  def download_tar
+    path = params[:path]&.split("/")&.drop(2)&.join("/")
+    path = CGI.unescape(path)
+    absolute_path = check_path_exist(path).to_s
+    return unless check_instructor(absolute_path)
+
+    if File.directory?(absolute_path)
+      tar_stream = StringIO.new("")
+      Gem::Package::TarWriter.new(tar_stream) do |tar|
+        Dir[File.join(absolute_path, '**', '**')].each do |file|
+          mode = File.stat(file).mode
+          relative_path = file.sub(%r{^#{Regexp.escape(absolute_path)}/?}, '')
+          if File.directory?(file)
+            tar.mkdir relative_path, mode
+          else
+            tar.add_file relative_path, mode do |tar_file|
+              File.open(file, "rb") { |f| tar_file.write f.read }
+            end
+          end
+        end
+      end
+      tar_stream.rewind
+      tar_stream.close
+      send_data tar_stream.string.force_encoding("binary"),
+                filename: "file_manager.tar",
+                type: "application/x-tar",
+                disposition: "attachment"
+    else
+      send_file(absolute_path,
+                filename: File.basename(absolute_path),
+                disposition: 'attachment')
+    end
+  end
+
+  def upload_file(path)
+    absolute_path = check_path_exist(path)
+    if absolute_path == BASE_DIRECTORY
+      flash[:error] = "You cannot upload files/create folders in the root directory " \
+         "#{view_context.link_to 'here', new_course_url, method: 'get'}" \
+         " if you want to create a new course."
+      flash[:html_safe] = true
+    else
+      raise ActionController::ForbiddenError unless File.directory?(absolute_path)
+
+      if check_instructor(absolute_path) && !params[:name].nil?
+        if params[:name] != ""
+          # Creating a folder
+          dir = "#{absolute_path}/#{params[:name]}"
+          FileUtils.mkdir_p(dir) unless File.directory?(dir)
+          flash[:success] = "Successfully created folder"
```

Comment:
When creating a folder with a name that already exists, the message "Successfully created folder" is still displayed.  The folder contents aren't affected, but this is probably not desirable